### DISCUSSION TITLE
Compile custom_target only when installing

### DISCUSSION THREAD

**Discussion by bruchar1** [mesonbuild/meson#11397] [Q&A]
Suppose I have something like this:

```
install_data = custom_target(
    mytarget,
    ...,
    build_by_default: false
)

meson.add_install_script(myscript, install_data, install_tag: 'mytag')
```

Is it possible to build the custom target only when installing, or when explicitly building it, but not when I compile the whole project? It seems the `build_by_default` behavior is overrided by the dependency on the install script...

**Comment by xclaesse** [comment]
`build_by_default` is indeed overridden when the target is used in an install script. It is an hugly hack that we should fix properly somehow, the relevant code is there: https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreter/mesonmain.py#L127

**Reply by bruchar1** [reply]
Here is my proposal. I'm not sure it's optimal, but I write it to start the discussion...

The interpreter could record a dict of `{target_output: target_name}` for the custom script, and keep it somewhere (into the `ExecutableSerialisation`?) It should probably only keep the target with `build_by_default: false`, as other targets will be already build.

Then, in the install process, in `run_install_script`, just before invoking the executable, it could check if each previously recorded target_output exists. If not, it would invoke `ninja` on the associated target name.

Remaining questions:
- would that mix of build and install processes be a problem? Otherwise, it would need to scan install targets twice: once for building missing targets, and once for performing the install.
- should it build the dependent target if `no_rebuild` is `true`?
- what about msvc backend? (but it is not supported by rebuild_all anyway...)

**Comment by eli-schwartz** [comment]
I am genuinely curious what the goal is of marking something as not built when building "all"? How do you separate the compile stage from the install stage? e.g. `ninja && sudo meson install`?

**Reply by bruchar1** [reply]
This is because this peticular custom_target is only useful for the installation step. When we build **all**, the goal is to ensure that everything is compiling. We dont want to spend time in the generation of installation data at this step.

---

We have a monorepo, with several libs and executables. From those libs and executables, we need to build an installation template for a _product_: i.e. copy all the needed files are the right location. The same shared lib can potentially be copied into different product templates. Therefore, I cannot simply use the installation prefix for that.

I wrote a script that takes the targets ( shared libs and exe) I want to bundle as a input, and it writes a file with the list of dll, lib, pdb files needed for that target (from the dependencies). Those shared libs and executables are not necessarily dependent of each other: it could be plugins, utilities commands, etc. Having a custom_target allows to compile all those related targets as one target.

A second script use the generated file list to copy them at the right location. This is the install step, performed using an install script. This step is tagged with the product name.

When I'm developing, I can build my lib using `meson compile mylib` (or `ninja compile mylib`).

When I'm creating a release, I can create my installation template using `meson install --tags myproduct`, and it will compile all (but only) the needed dependencies.

I hope this makes sense.

**Reply by bruchar1** [reply]
Rethinking about that... I should probably replace the custom_target with a target_alias, and move the dependency finder to the install script....

However, having `build_by_default` overrided by the install script dependency is still a bug imho...

**Reply by xclaesse** [reply]
> However, having `build_by_default` overrided by the install script dependency is still a bug imho...

Probably fair to at least warn in that case, indeed.
